src/ProjectOrigin.Registry.IntegrationTests/ThroughputTests.cs

```diff
@@ -0,0 +1,163 @@
+            .WithDockerfile("ProjectOrigin.Registry.Server/Dockerfile")
+            .Build();
+
+        _registryContainer = new Lazy<IContainer>(() =>
+        {
+            var verifierUrl = $"http://{_verifierContainer.IpAddress}:{GrpcPort}";
+            return new ContainerBuilder()
+                .WithImage(_registryImage)
+                .WithPortBinding(GrpcPort, true)
+                .WithEnvironment("Verifiers__project_origin.electricity.v1", verifierUrl)
+                .WithEnvironment("RegistryName", RegistryName)
+                .WithEnvironment("IMMUTABLELOG__TYPE", "log")
+                .WithEnvironment("VERIFIABLEEVENTSTORE__BATCHSIZEEXPONENT", "0")
+                .WithWaitStrategy(
+                    Wait.ForUnixContainer()
+                        .UntilPortIsAvailable(GrpcPort)
+                    )
+                .Build();
+        });
+
+    }
+
+    private GrpcChannel CreateRegistryChannel()
+    {
+        return GrpcChannel.ForAddress($"http://{_registryContainer.Value.IpAddress}:{GrpcPort}");
+    }
+
+    [Fact(Skip = "Manual test")]
+    public async Task TestThroughput()
+    {
+        int concurrency = 10;
+        int concurrentRequests = 500;
+        var tasks = new Task[concurrency];
+        var stopwatch = new Stopwatch();
+
+        ConcurrentQueue<Registry.V1.Transaction> queued = new ConcurrentQueue<Registry.V1.Transaction>();
+        ConcurrentBag<Registry.V1.Transaction> completed = new ConcurrentBag<Registry.V1.Transaction>();
+        using var channel = CreateRegistryChannel();
+
+        stopwatch.Start();
+        for (int i = 0; i < concurrency; i++)
+        {
+            tasks[i] = Task.Run(async () =>
+            {
+                while (true)
+                {
+                    if (queued.Count < concurrentRequests)
+                    {
+                        Console.WriteLine("sending request");
+                        queued.Enqueue(await SendRequest(channel).ConfigureAwait(false));
+                    }
+                    else
+                    {
+                        if (queued.TryDequeue(out var transaction))
+                        {
+                            if (await IsCommitted(channel, transaction))
+                            {
+                                Console.WriteLine($"completed request {completed.Count}");
+                                completed.Add(transaction);
+                            }
+                            else
+                            {
+                                Console.WriteLine($"re-queueing request {completed.Count}");
+                                queued.Enqueue(transaction);
+                                await Task.Delay(25).ConfigureAwait(false);
+                            }
+                        }
+                    }
+                }
+            });
+        }
+        await Task.Delay(TimeSpan.FromMinutes(5)).ConfigureAwait(false);
```

Comment:
Idea: You could use a CancellationToken and then request cancellation when everything is processed

Reply:
As it is designed now, it doesn't complete. it starts 10 threads, that allows to concurrenly have 500 transactions in processing on the registry, and simple measures the amounts of transactions completed within 5 minutes to test the throughput.